src/containers/permissions/DataSetColumnPermissionsSection.js

```diff
@@ -0,0 +1,263 @@
+
+const ToggleWrapper = styled.div`
+  align-items: center;
+  display: flex;
+  min-height: 47px; /* to compensate for both spinner and icon */
+  justify-content: center;
+  margin-right: -10px; /* for toggle icon alignment with checkboxes */
+  min-width: 47px; /* to compensate for both spinner and icon */
+`;
+
+const DataSetColumnPermissionsSection = ({
+  dataSetColumns,
+  objectKey,
+  permissions,
+  permissionType,
+  principal,
+} :{|
+  dataSetColumns :List<Map<FQN, List>>;
+  objectKey :List<UUID>;
+  permissions :Map<List<UUID>, Ace>;
+  permissionType :PermissionType;
+  principal :Principal;
+|}) => {
+
+  const [isPermissionAssignedToAll, setIsPermissionAssignedToAll] = useState(false);
+  const [isPermissionAssignedToOnlyNonPII, setIsPermissionAssignedToOnlyNonPII] = useState(false);
+  const [selectedToggle, setSelectedToggle] = useState('');
+
+  const myKeys :Set<List<UUID>> = useSelector(selectMyKeys());
+  const aceForUpdate = (new AceBuilder()).setPermissions([permissionType]).setPrincipal(principal).build();
+
+  const dispatch = useDispatch();
+
+  const columnIds :List<UUID> = useMemo(() => (
+    dataSetColumns.map((column :Map<FQN, List>) => getPropertyValue(column, [FQNS.OL_ID, 0]))
+  ), [dataSetColumns]);
+  const maybePropertyTypes :Map<UUID, PropertyType> = useSelector(selectPropertyTypes(columnIds));
+  const propertyTypesHash :number = maybePropertyTypes.hashCode();
+
+  useEffect(() => {
+    const dataSetId :UUID = objectKey.get(0, '');
+    const { isAssignedToAll, isAssignedToOnlyNonPII } = computePermissionAssignments(
+      myKeys,
+      dataSetColumns,
+      dataSetId,
+      permissions,
+      permissionType,
+      maybePropertyTypes,
+    );
+    setIsPermissionAssignedToAll(isAssignedToAll);
+    setIsPermissionAssignedToOnlyNonPII(isAssignedToOnlyNonPII);
+    /* eslint-disable-next-line */
+  }, [
+    dataSetColumns,
+    myKeys,
+    objectKey,
+    permissions,
+    permissionType,
+    propertyTypesHash,
+  ]);
+
+  const updatePermissionsRS :?RequestState = useRequestState([PERMISSIONS, UPDATE_PERMISSIONS]);
+
+  const togglePermissionAssignmentToAll = () => {
+    if (!isPending(updatePermissionsRS)) {
+      const permissionsToUpdate = Map().withMutations((mutator) => {
+        dataSetColumns.forEach((column :Map<FQN, List>) => {
+          const columnId :UUID = getPropertyValue(column, [FQNS.OL_ID, 0]);
+          const key :List<UUID> = List([objectKey.get(0), columnId]);
+          if (myKeys.has(key)) {
+            mutator.set(key, aceForUpdate);
+          }
+        });
+      });
+      dispatch(
+        updatePermissions([{
+          actionType: isPermissionAssignedToAll ? ActionTypes.REMOVE : ActionTypes.ADD,
+          permissions: permissionsToUpdate
+        }])
+      );
+      setIsPermissionAssignedToAll(!isPermissionAssignedToAll);
+    }
+  };
+
+  const togglePermissionAssignmentToOnlyNonPII = () => {
```

Comment:
How resilient are these toggle functions to a user rage clicking the action? I see that nothing happens if it's pending, but just curious.

Reply:
good question... i would say they are actually pretty resilient:
![e3b9111ba51e4f77e253e5363ca3f66e](https://user-images.githubusercontent.com/32921059/116613152-32ec6200-a8ed-11eb-9e6c-df920c38c3f7.gif)